internal/controller/controllers/cluster_controller.go

```diff
@@ -171,17 +228,121 @@ func (r *ClusterReconciler) updateStatus(ctx context.Context, cluster *adiiov1al
+	r.syncClusterState(cluster, c)
 
 	if err := r.Status().Update(ctx, cluster); err != nil {
 		r.Log.WithError(err).Error("failed to update error status")
-		return ctrl.Result{}, errors.Wrap(err, "failed to create cluster")
+		return ctrl.Result{Requeue: true}, nil
 	}
 
-	return ctrl.Result{}, nil
+	r.Log.Infof("Updated cluster %s %s state", cluster.Name, cluster.Namespace)
+	return ctrl.Result{RequeueAfter: defaultRequeueClusterDuration}, nil
+}
+
+func (r *ClusterReconciler) updateIfNeeded(ctx context.Context, cluster *adiiov1alpha1.Cluster, c *common.Cluster) (bool, ctrl.Result, error) {
+	update := false
+
+	params := &models.ClusterUpdateParams{}
+
+	spec := cluster.Spec
+
+	updateString := func(new, old string, target **string) {
+		if new != old {
+			*target = swag.String(new)
+			update = true
+			r.Log.Infof("updating %s != %s", old, new)
+		}
+	}
+
+	updateString(spec.Name, c.Name, &params.Name)
+
+	if spec.OpenshiftVersion != c.OpenshiftVersion {
+		return false, ctrl.Result{}, errors.Errorf("Openshift version cannot be updated")
+	}
+
+	updateString(spec.BaseDNSDomain, c.BaseDNSDomain, &params.BaseDNSDomain)
+	updateString(spec.ClusterNetworkCidr, c.ClusterNetworkCidr, &params.ClusterNetworkCidr)
+
+	if spec.ClusterNetworkHostPrefix != c.ClusterNetworkHostPrefix {
+		params.ClusterNetworkHostPrefix = swag.Int64(spec.ClusterNetworkHostPrefix)
+		update = true
+	}
+
+	updateString(spec.ServiceNetworkCidr, c.ServiceNetworkCidr, &params.ServiceNetworkCidr)
+	updateString(spec.APIVip, c.APIVip, &params.APIVip)
+	updateString(spec.APIVipDNSName, swag.StringValue(c.APIVipDNSName), &params.APIVipDNSName)
+	updateString(spec.IngressVip, c.IngressVip, &params.IngressVip)
+	updateString(spec.MachineNetworkCidr, c.MachineNetworkCidr, &params.MachineNetworkCidr)
+	updateString(spec.SSHPublicKey, c.SSHPublicKey, &params.SSHPublicKey)
+
+	if spec.VIPDhcpAllocation != swag.BoolValue(c.VipDhcpAllocation) {
+		params.VipDhcpAllocation = swag.Bool(spec.VIPDhcpAllocation)
+	}
+
+	updateString(spec.HTTPProxy, c.HTTPProxy, &params.HTTPProxy)
+	updateString(spec.HTTPSProxy, c.HTTPSProxy, &params.HTTPSProxy)
+	updateString(spec.NoProxy, c.NoProxy, &params.NoProxy)
+
+	if spec.UserManagedNetworking != swag.BoolValue(c.UserManagedNetworking) {
+		params.UserManagedNetworking = swag.Bool(spec.UserManagedNetworking)
+	}
+
+	updateString(spec.AdditionalNtpSource, c.AdditionalNtpSource, &params.AdditionalNtpSource)
+
+	// TODO: handle InstallConfigOverrides
+
+	data, err := r.getPullSecret(ctx, spec.PullSecretRef.Name, spec.PullSecretRef.Namespace)
+	if err != nil {
+		return false, ctrl.Result{}, errors.Wrap(err, "failed to get pull secret for update")
+	}
+	updateString(data, c.PullSecret, &params.PullSecret)
+
+	if !update {
+		return update, ctrl.Result{}, nil
+	}
+
+	updatedCluster, err := r.Installer.UpdateClusterInternal(ctx, installer.UpdateClusterParams{
```

Comment:
don't we want to check if client error so we could requeue only if it doesn't?

Reply:
Not sure i understand.

Reply:
Because the function return an error the reply will be given by updateState  so it will be requeued.